src/type-logic.ts

```diff
@@ -0,0 +1,576 @@
+        return { kind: 'concrete real', value: text };
+      }
+      if (text.startsWith('*') && text.endsWith('*<sub>𝔽</sub>')) {
+        const innerText = text.slice(1, -14);
+        let value;
+        if (innerText === '-0') {
+          value = -0;
+        } else if (innerText === '+&infin;') {
+          value = 2e308;
+        } else if (innerText === '-&infin;') {
+          value = -2e308;
+        } else {
+          value = parseFloat(innerText);
+        }
+        return { kind: 'concrete number', value };
+      }
+      if (text === '*NaN*') {
+        return { kind: 'concrete number', value: 0 / 0 };
+      }
+      if (text.startsWith('*') && text.endsWith('*<sub>ℤ</sub>')) {
+        return { kind: 'concrete bigint', value: BigInt(text.slice(1, -14)) };
+      }
+      if (text === 'an ECMAScript language value' || text === 'ECMAScript language values') {
+        return { kind: 'ES value' };
+      }
+      if (text === 'a String' || text === 'Strings') {
+        return { kind: 'string' };
+      }
+      if (text === 'a Number' || text === 'Numbers') {
+        return { kind: 'number' };
+      }
+      if (text === 'a Boolean' || text === 'Booleans') {
+        return { kind: 'boolean' };
+      }
+      if (text === 'a BigInt' || text === 'BigInts') {
+        return { kind: 'bigint' };
+      }
+      if (text === 'an integral Number' || text === 'integral Numbers') {
+        return { kind: 'integral number' };
+      }
+      if (text === 'a mathematical value' || text === 'mathematical values') {
+        return { kind: 'real' };
+      }
+      if (text === 'an integer' || text === 'integers') {
+        return { kind: 'integer' };
+      }
+      if (text === 'a non-negative integer' || text === 'non-negative integers') {
+        return { kind: 'non-negative integer' };
+      }
+      if (text === 'a negative integer' || text === 'negative integers') {
+        return { kind: 'negative integer' };
+      }
+      if (text === 'a positive integer' || text === 'positive integers') {
+        return { kind: 'positive integer' };
+      }
+      if (text === 'a time value' || text === 'time values') {
+        return {
+          kind: 'union',
+          of: [{ kind: 'integral number' }, { kind: 'concrete number', value: 0 / 0 }],
+        };
+      }
+      if (text === '*null*') {
+        return { kind: 'null' };
+      }
+      if (text === '*undefined*') {
+        return { kind: 'undefined' };
+      }
+      break;
+    }
+    case 'unused': {
+      // this is really only a return type, but might as well handle it
+      return { kind: 'enum value', value: '~unused~' };
+    }
+  }
+  return { kind: 'unknown' };
+}
+
+export function isCompletion(
+  type: Type,
+): type is Type & { kind: 'normal completion' | 'abrupt completion' | 'union' } {
+  return (
+    type.kind === 'normal completion' ||
+    type.kind === 'abrupt completion' ||
+    (type.kind === 'union' && type.of.some(isCompletion))
```

Comment:
You can't mix completions and non-completions, right? So shouldn't this just check `isCompletion(type.of[0])`?

Reply:
Not syntactically, but it's possible for it to arise as a result of (more sophisticated than currently implemented) type analysis. For example, consider

```
1. Let _x_ be Completion(_something_).
1. If _x_ is a normal completion], then
  1. Set _x_ to ! _x_.
1. Return _x_.
```

On the last line, if we tracked the types of variables, we might think `_x_` is either a completion or not. The user shouldn't write this program but the analysis shouldn't go off the rails if they do.

Reply:
Ah right, we arrive at this through type analysis, not just type declarations. 👍